### PR TITLE
Add standard-asset-apis package with routing support

### PR DESCRIPTION
**Pull request by peterdemartini**
- [x] Add the basics of a standard-asset-api package
- [x] Add `KeyRouter`
- [x] Add `DateRouter`
- [x] Add `HashRouter`
- [x] Add `FieldRouter`
- [x] Add `debug_routes` processor 

BREAKING CHANGES:

- Change `buckets` to `partitions`
- Change default id field to `_key` from `_id`

### CODE REVIEW

**Comment by peterdemartini** [comment]
I am going to start manually testing this to make sure I didn't break anything.

@ciorg @jsnoble would you mind taking a look at it in the mean time?

**Comment by peterdemartini** [comment]
I ran some tests and this looks good so far

**Review by jsnoble — APPROVED** [review]
LGTM

File: packages/standard-asset-apis/src/routers/DateRouter.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+export interface DateRouterConfig {
+    /**
+     * This is the primary date field on the record
+    */
+    field: string;
+
+    /**
+     * This is used determine how specific the date
+     * the date routing should be
```
Comment (ciorg) [inline packages/standard-asset-apis/src/routers/DateRouter.ts:134]: typo - repeat of `the date`

File: packages/standard-asset-apis/src/routers/HashRouter.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    lookup(record: DataEntity): string {
+        const bucket = fnv1a(this.getHash(record)) % this.partitions;
+        return bucket.toString();
+    }
+}
+
+export interface HashRouterConfig {
+    /**
+     * The number of partitions to created based of the hash
```
Comment (ciorg) [inline packages/standard-asset-apis/src/routers/HashRouter.ts:41]: typo `to be created...`

File: packages/standard-asset-apis/src/routers/HashRouter.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    */
+    partitions: number;
+
+    /**
+     * This fields to use for the key, this must
```
Comment (ciorg) [inline packages/standard-asset-apis/src/routers/HashRouter.ts:46]: typo - The fields, or just fields

File: packages/standard-asset-apis/src/routers/interfaces.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    DATA = 'DATA',
+    STORAGE = 'STORAGE',
+}
+
+export interface Router {
+    readonly kind: RouterKind;
+
+    /**
+     * When called this uses the routing algorithm to return the
+     * route to use, which usually string|number with meaning to the
```
Comment (ciorg) [inline packages/standard-asset-apis/src/routers/interfaces.ts:13]: typo. - `which usually is a ...`

File: test/hash_router/schema-spec.ts [hunk 5]
```diff
@@ -26,23 +28,23 @@ describe('Hash Router Schema', () => {
...
+            await expect(makeSchema({ fields: undefined, partitions: 1 })).toResolve();
+            await expect(makeSchema({ fields: JSON.stringify('this ia a string') as any, partitions: 1 })).toReject();
+            await expect(makeSchema({ fields: 42 as any, partitions: 1 })).toReject();
+            await expect(makeSchema({ fields: [42] as any, partitions: 1 })).toReject();
+
+            await expect(makeSchema({ fields: [], partitions: 1 })).toResolve();
+            await expect(makeSchema({ fields: ['someField'], partitions: 1 })).toResolve();
         });
 
         it('should throw an error if `buckets` is not a positive number', async () => {
```
Comment (ciorg) [inline test/hash_router/schema-spec.ts:41]: should be `partitions` instead of buckets

**Review by ciorg — COMMENTED** [review]
just a few typos otherwise LGTM

**Pull request merged**